src/npctalk.cpp

```diff
@@ -2694,38 +2694,139 @@ void talk_effect_fun_t::set_location_variable( const JsonObject &jo, const std::
     }
 
+    std::optional<str_or_var> search_target;
+    std::optional<std::string> search_type;
+    dbl_or_var dov_target_min_radius = get_dbl_or_var( jo, "target_min_radius", false, 0 );
+    dbl_or_var dov_target_max_radius = get_dbl_or_var( jo, "target_max_radius", false, 0 );
+    int target_types = 0;
+    if( jo.has_member( "terrain" ) ) {
+        target_types++;
+        search_type = "terrain";
+    }
+    if( jo.has_member( "furniture" ) ) {
+        target_types++;
+        search_type = "furniture";
+    }
+    if( jo.has_member( "monster" ) ) {
+        target_types++;
+        search_type = "monster";
+    }
+    if( jo.has_member( "npc" ) ) {
+        target_types++;
+        search_type = "npc";
+    }
+    if( jo.has_member( "trap" ) ) {
+        target_types++;
+        search_type = "trap";
+    }
+    if( target_types == 1 ) {
+        search_target = get_str_or_var( jo.get_member( search_type.value() ), search_type.value(), true );
+    } else if( target_types > 1 ) {
+        jo.throw_error( "Can only have one of terrain, furniture, monster, trap, or npc." );
+    }
+
     var_info var = read_var_info( jo.get_object( member ) );
     var_type type = var.type;
     std::string var_name = var.name;
 
     std::vector<effect_on_condition_id> true_eocs = load_eoc_vector( jo, "true_eocs" );
     std::vector<effect_on_condition_id> false_eocs = load_eoc_vector( jo, "false_eocs" );
 
-    function = [dov_min_radius, dov_max_radius, var_name, outdoor_only, target_params,
-                                is_npc, type, dov_x_adjust, dov_y_adjust, dov_z_adjust, z_override, true_eocs,
-                    false_eocs]( dialogue const & d ) {
+    function = [dov_min_radius, dov_max_radius, var_name, outdoor_only, target_params, is_npc, type,
+                                dov_x_adjust, dov_y_adjust, dov_z_adjust, z_override, true_eocs, false_eocs, search_target,
+                    search_type, dov_target_min_radius, dov_target_max_radius]( dialogue const & d ) {
+        map &here = get_map();
+        tripoint_abs_ms avatar_pos = get_avatar().get_location();
         talker *target = d.actor( is_npc );
-        tripoint talker_pos = get_map().getabs( target->pos() );
+        tripoint talker_pos = here.getabs( target->pos() );
         tripoint target_pos = talker_pos;
-
-        int max_radius = dov_max_radius.evaluate( d );
         if( target_params.has_value() ) {
             const tripoint_abs_omt omt_pos = mission_util::get_om_terrain_pos( target_params.value() );
             target_pos = tripoint( project_to<coords::ms>( omt_pos ).x(), project_to<coords::ms>( omt_pos ).y(),
                                    project_to<coords::ms>( omt_pos ).z() );
-        } else if( max_radius > 0 ) {
+        }
+        const tripoint_abs_ms abs_ms( target_pos );
+        bool shifted = false;
+        if( !here.inbounds( abs_ms ) ) {
+            g->place_player_overmap( project_to<coords::omt>( abs_ms ), false );
```

Comment:
Load a (tiny)map and check there please. Don't move the avatar around as it causes no end of trouble (ex: #60096 and the myriad of issues in the TCL).

You might have to manually spawn and unload monsters though.